packages/teuchos/parameterlist/src/Teuchos_ParameterListModifier.cpp

```diff
@@ -0,0 +1,216 @@
+// @HEADER
+// ***********************************************************************
+//
+//                    Teuchos: Common Tools Package
+//                 Copyright (2004) Sandia Corporation
+//
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+// @HEADER
+
+
+#include "Teuchos_ParameterListModifier.hpp"
+#include "Teuchos_ParameterList.hpp"
+#include "Teuchos_StrUtils.hpp"
+
+namespace Teuchos {
+
+
+// Constructors
+
+
+ParameterListModifier::ParameterListModifier()
+  :name_("ANONYMOUS")
+{}
+
+
+ParameterListModifier::ParameterListModifier(const std::string &name_in)
+  :name_(name_in)
+{}
+
+void ParameterListModifier::printDoc(std::string const& docString, std::ostream &out) const{
+  StrUtils::printLines(out,"# ",docString);
+  out << "#  Modifier Used: " << name_ << std::endl;
+}
+
+
+Array<std::string> ParameterListModifier::findMatchingBaseNames(const ParameterList &pl,
+    const std::string &base_name, const bool &find_parameters, const bool &find_sublists) const{
+  Array<std::string> matches(0);
+  ParameterList::ConstIterator itr;
+  for (itr = pl.begin(); itr != pl.end(); ++itr) {
+    const std::string &name = pl.name(itr);
```

Comment:
If `name` has an overload that takes a parameter list entry instead of an iterator, then we can use range for here.

Reply:
I used an iterator here to get the name without the full path.  I think a lot of people are surprised when they start working with nested parameters/sublists and discover that the `updateSubListNames` method changes `name_` to be the full path.  Something like a `shortName` method would be awfully handy and make using range easier.  I'll add it as an additional commit with the option of dropping it if you have a better idea.

Reply:
@danielsjensen1 I was mainly just curious if `name` had that overload -- if it's trouble, please don't worry about it.  Thanks!

Reply:
Actually, while digging into this more I think we need to open a separate issue for `name`.  The following code shows why `name` and `setName` are so confusing and arguably not working correctly:
```
  ParameterList foo("foo");
  foo.sublist("A").sublist("SubA");
  foo.print();
  foo.sublist("A").sublist("SubA").setName("New SubA");
  std::cout << "foo.sublist(""A"").sublist(""SubA"").name() = " << foo.sublist("A").sublist("SubA").name() << std::endl;
  foo.print();
  foo.sublist("A").sublist("New SubA", true);
```
Output:
```
 A -> 
  SubA -> 
   [empty list]
foo.sublist(A).sublist(SubA).name() = New SubA
 A -> 
  SubA -> 
   [empty list]
 
 
 p=0: *** Caught standard std::exception of type 'Teuchos::Exceptions::InvalidParameterName' :
 
  The sublist foo->A->"New SubA" does not exist!
  
  Throw number = 1
```
It appears that you can't really change the name of a sublist without removing it from the parameter list and reattaching it with a new name.  I've removed my changes to `name` and `setName` for now since this isn't an easy fix.